2011-08-24  Andreas Schwab  <schwab@redhat.com>

	* elf/dl-close.c (_dl_close_worker): Reset private search list if
	it wasn't used.

---
 elf/dl-close.c |    7 +++++++
 1 file changed, 7 insertions(+)

--- a/elf/dl-close.c
+++ b/elf/dl-close.c
@@ -421,6 +421,13 @@
 
 	      imap->l_scope_max = new_size;
 	    }
+	  else if (new_list != NULL)
+	    {
+	      /* We didn't change the scope array, so reset the search
+		 list.  */
+	      imap->l_searchlist.r_list = NULL;
+	      imap->l_searchlist.r_nlist = 0;
+	    }
 
 	  /* The loader is gone, so mark the object as not having one.
 	     Note: l_idx != IDX_STILL_USED -> object will be removed.  */
